tls/s2n_config.c

```diff
@@ -683,6 +724,136 @@ int s2n_config_get_cert_type(struct s2n_config *config, s2n_cert_type *cert_type
+                                     uint8_t *key, uint32_t key_len,
+                                     uint64_t intro_time_in_seconds_from_epoch)
+{
+    notnull_check(config);
+    notnull_check(name);
+    notnull_check(key);
+
+    if (!config->use_tickets) {
+        return 0;
+    }
+
+    GUARD(s2n_config_wipe_expired_ticket_crypto_keys(config, -1));
+
+    S2N_ERROR_IF(key_len == 0, S2N_ERR_INVALID_TICKET_KEY_LENGTH);
+
+    S2N_ERROR_IF(s2n_find_ticket_key(config, name) || name_len == 0 || name_len > S2N_TICKET_KEY_NAME_LEN, S2N_ERR_INVALID_TICKET_KEY_NAME_OR_NAME_LENGTH);
+
+    S2N_ERROR_IF(config->ticket_keys->num_of_elements >= S2N_MAX_TICKET_KEYS, S2N_ERR_TICKET_KEY_LIMIT);
+
+    uint64_t now;
+    uint16_t insert_index = 0;
+    uint8_t output_pad[S2N_AES256_KEY_LEN + S2N_TICKET_AAD_IMPLICIT_LEN];
+    struct s2n_blob out_key = { .data = output_pad, .size = sizeof(output_pad) };
+    struct s2n_blob in_key = { .data = key, .size = key_len };
+    struct s2n_blob salt = { .size = 0 };
+    struct s2n_blob info = { .size = 0 };
+
+    struct s2n_blob allocator;
+    struct s2n_ticket_key *session_ticket_key;
+
+    GUARD(s2n_alloc(&allocator, sizeof(struct s2n_ticket_key)));
+    session_ticket_key = (struct s2n_ticket_key *) (void *) allocator.data;
+
+    struct s2n_hmac_state hmac;
+
+    GUARD(s2n_hmac_new(&hmac));
+    GUARD(s2n_hkdf(&hmac, S2N_HMAC_SHA256, &salt, &in_key, &info, &out_key));
+    GUARD(s2n_hmac_free(&hmac));
+
+    struct s2n_hash_state hash = { 0 };
+    uint8_t hash_output[SHA_DIGEST_LENGTH];
+
+    GUARD(s2n_hash_new(&hash));
+    GUARD(s2n_hash_init(&hash, S2N_HASH_SHA1));
+    GUARD(s2n_hash_update(&hash, out_key.data, out_key.size));
+    GUARD(s2n_hash_digest(&hash, hash_output, SHA_DIGEST_LENGTH));
+
+    if (config->ticket_key_hashes->num_of_elements >= S2N_MAX_TICKET_KEY_HASHES) {
+        GUARD(s2n_array_free(config->ticket_key_hashes));
+        notnull_check(config->ticket_key_hashes = s2n_array_new(SHA_DIGEST_LENGTH));
+    }
+
+    GUARD(s2n_verify_unique_ticket_key(config, hash_output, &insert_index));
+
+    /* Insert hash key into a sorted array at known index */
+    struct uint8_t *hash_element = s2n_array_insert(config->ticket_key_hashes, insert_index);
+    memcpy_check(hash_element, hash_output, SHA_DIGEST_LENGTH);
+
+    memcpy_check(session_ticket_key->key_name, name, S2N_TICKET_KEY_NAME_LEN);
+    memcpy_check(session_ticket_key->aes_key, out_key.data, S2N_AES256_KEY_LEN);
+    out_key.data = output_pad + S2N_AES256_KEY_LEN;
+    memcpy_check(session_ticket_key->implicit_aad, out_key.data, S2N_TICKET_AAD_IMPLICIT_LEN);
+
+    if (intro_time_in_seconds_from_epoch == 0) {
+        GUARD(config->monotonic_clock(config->monotonic_clock_ctx, &now));
+        session_ticket_key->intro_timestamp = now;
+    } else {
+        session_ticket_key->intro_timestamp = (intro_time_in_seconds_from_epoch * ONE_SEC_IN_NANOS);
+    }
+
+    /* Keys are stored from oldest to newest */
```

Comment:
this isn't strcitly true, as you could add keys out of order with different intro times, right?

Reply:
Oh yeah. Forgot to update this when I added intro_time to the API. Will fix it.